File: scripts/needo-roadshow/make-contact-sheet.mjs
import fs from "node:fs";
import path from "node:path";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";

const require = createRequire(import.meta.url);
const sharp = require("sharp");
const here = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(here, "../..");
const previewDir = path.join(root, "outputs/needo-roadshow-2026-08-23/previews");
const output = path.join(root, "outputs/needo-roadshow-2026-08-23/contact-sheet.png");
const files = fs.readdirSync(previewDir)
  .filter((file) => /^slide-\d+\.png$/.test(file))
  .sort((a, b) => Number(a.match(/\d+/)[0]) - Number(b.match(/\d+/)[0]));

const cols = 4;
const thumbW = 480;
const thumbH = 270;
const gap = 24;
const labelH = 32;
const rows = Math.ceil(files.length / cols);
const width = cols * thumbW + (cols + 1) * gap;
const height = rows * (thumbH + labelH) + (rows + 1) * gap;
const composites = [];

for (let i = 0; i < files.length; i += 1) {
  const col = i % cols;
  const row = Math.floor(i / cols);
  const left = gap + col * (thumbW + gap);
  const top = gap + row * (thumbH + labelH + gap);
  const image = await sharp(path.join(previewDir, files[i])).resize(thumbW, thumbH, { fit: "fill" }).png().toBuffer();
  composites.push({ input: image, left, top });
  const labelSvg = Buffer.from(`<svg xmlns="http://www.w3.org/2000/svg" width="${thumbW}" height="${labelH}"><rect width="100%" height="100%" fill="#FFFFFF"/><text x="10" y="22" font-family="Arial" font-size="17" font-weight="700" fill="#2F5F4A">${String(i + 1).padStart(2, "0")}</text></svg>`);
  composites.push({ input: labelSvg, left, top: top + thumbH });
}

await sharp({ create: { width, height, channels: 3, background: "#EDF5F1" } }).composite(composites).png().toFile(output);
console.log(output);

File: scripts/needo-roadshow-template/make_contact_sheet.py
#!/usr/bin/env python3
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

ROOT = Path(__file__).resolve().parents[2]
PREVIEWS = ROOT / "outputs/needo-roadshow-template-2026-08-23/previews"
OUTPUT = ROOT / "outputs/needo-roadshow-template-2026-08-23/contact-sheet-template.png"

files = sorted(PREVIEWS.glob("slide-*.png"), key=lambda path: int(path.stem.split("-")[-1]))
cols = 4
thumb_w, thumb_h = 480, 270
label_h, gap = 30, 20
rows = (len(files) + cols - 1) // cols
sheet = Image.new("RGB", (cols * thumb_w + (cols + 1) * gap, rows * (thumb_h + label_h) + (rows + 1) * gap), "#EEF1EA")
draw = ImageDraw.Draw(sheet)
font = ImageFont.load_default(size=18)

for index, file in enumerate(files):
    col, row = index % cols, index // cols
    x = gap + col * (thumb_w + gap)
    y = gap + row * (thumb_h + label_h + gap)
    image = Image.open(file).convert("RGB").resize((thumb_w, thumb_h), Image.Resampling.LANCZOS)
    sheet.paste(image, (x, y))
    draw.rectangle((x, y + thumb_h, x + thumb_w, y + thumb_h + label_h), fill="white")
    draw.text((x + 10, y + thumb_h + 5), f"{index + 1:02d}", fill="#455F51", font=font)

sheet.save(OUTPUT)
print(OUTPUT)
